Process #1878: Add guidelines to support 1.4 consistency review PRs (#2236)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: guidelines/contributors/editorial-guidelines.md

* **Entity Scope:** These rules apply to all entities defined in the FOCUS specification (e.g., Columns, Attributes, Datasets, Objects).

### Column Values

* **Inline Code for Values:** When referencing specific string values that a column can contain, the value MUST be enclosed in backticks to render as inline code (e.g., `Usage`). 

* **Consistent Formatting:** This formatting MUST be used consistently across all normative requirements, descriptions, and tables when referencing a specific string value.

### Comparison Terminology

To ensure consistent language when describing relationships and evaluations between values, normative requirements and standard prose MUST use the following terminology:

* **Numeric Values:** Use "equal" or "be equal to" when comparing numeric values, costs, quantities, or mathematical sums (e.g., "ContractedCost MUST be equal to BilledCost"). Do not use "match".
* **Identifiers and Strings:** Use "match" when comparing strings, IDs, or names (e.g., "HostProviderName MUST match ServiceProviderName"). Do not use "equal".

> * **State and Conditions:** Use "is" when evaluating if a column contains a specific value (e.g., "when ChargeCategory is `Purchase`"). Do not use "equals".

Author: @kk09v (cc @ijurica) 
In support of your work on #2175, what do you think about this set of guidelines?

Feel free to suggest an alternative, but I think it gets to the heart of what you were attempting to do.

Reviewer: Looks good.

Exceptions:

Other forms of "is" are used.
- 'ChargeCategory MAY equal "Usage" when the record aggregates charges across multiple allowed values other than "Tax" (for example, aggregation of "Usage" and "Credit" is allowed, but not "Usage" and "Tax").'
  - This was changed to 'MAY be "Usage"'

Inequalities would use the term "equal" even if they're on non-metrical fields.
- 'BillingPeriodLastUpdated MUST be greater than or equal to [BillingPeriodCreated](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/2175#datasets.billingperiod.billingperiodcreated).'

This one I don't have a rule for. I suggested changing it to "semantically equivalent" to not trip over the use of equal.

- "PricingUnit MUST be semantically equal to the corresponding pricing measurement unit provided in service-provider-published price list."

Author: Okay!  Then how about adding two more rules?

----

### Comparison Terminology

To ensure consistent language when describing relationships and evaluations between values, normative requirements and standard prose MUST use the following terminology:

* **Numeric Values:** Use "equal" or "be equal to" when comparing numeric values, costs, quantities, or mathematical sums (e.g., "ContractedCost MUST be equal to BilledCost"). Do not use "match".
* **Identifiers and Strings:** Use "match" when comparing strings, IDs, or names (e.g., "HostProviderName MUST match ServiceProviderName"). Do not use "equal".
* **State and Conditions:** Use forms of the verb "to be" (e.g., "is", "are", "be") when evaluating if a column contains a specific value or state (e.g., "when ChargeCategory is `Purchase`" or "ChargeCategory MAY be `Usage`"). Do not use "equals" or "equal".
* **Inequalities:** When expressing range or directional limits, the standard phrases "greater than or equal to" and "less than or equal to" MUST be used, even for non-numeric data types like timestamps or dates (e.g., "BillingPeriodLastUpdated MUST be greater than or equal to BillingPeriodCreated"). 
* **Semantic Comparisons:** When comparing concepts, formats, or values that share the same meaning but may not be strictly identical strings, use "equivalent" or "semantically equivalent" (e.g., "PricingUnit MUST be semantically equivalent to the corresponding pricing measurement unit..."). Do not use "equal".

```suggestion
### Comparison Terminology

To ensure consistent language when describing relationships and evaluations between values, normative requirements and standard prose MUST use the following terminology:

* **Numeric Values:** Use "equal" or "be equal to" when comparing numeric values, costs, quantities, or mathematical sums (e.g., "ContractedCost MUST be equal to BilledCost"). Do not use "match".
* **Identifiers and Strings:** Use "match" when comparing strings, IDs, or names (e.g., "HostProviderName MUST match ServiceProviderName"). Do not use "equal".
* **State and Conditions:** Use forms of the verb "to be" (e.g., "is", "are", "be") when evaluating if a column contains a specific value or state (e.g., "when ChargeCategory is `Purchase`" or "ChargeCategory MAY be `Usage`"). Do not use "equals" or "equal".
* **Inequalities:** When expressing range or directional limits, the standard phrases "greater than or equal to" and "less than or equal to" MUST be used, even for non-numeric data types like timestamps or dates (e.g., "BillingPeriodLastUpdated MUST be greater than or equal to BillingPeriodCreated"). 
* **Semantic Comparisons:** When comparing concepts, formats, or values that share the same meaning but may not be strictly identical strings, use "equivalent" or "semantically equivalent" (e.g., "PricingUnit MUST be semantically equivalent to the corresponding pricing measurement unit..."). Do not use "equal".
```

Author: I'm committing this.  We can review further in #2111 once it's merged in.

* **Example** (Markdown, illustrative):

  ```md

>   * ... MUST be null when ChargeCategory is `Tax`.

Reviewer: Interesting. I know we talked about this during maintainers meeting, but I had thought about it in the context of tables, which is what we were looking at. Seeing it used in text areas feels a lot different.
Also, if we change defined values - which are strings - to use graves instead of quotes, how are we going to differentiate JSON `true` value from the string "true" ?

Reviewer 2: I pulled back the PR for double quotes in any case. Perhaps this is something to be discussed for a future verrsion.

Author: @kk09v We were explicit in the Maintainers meeting that this would be a change to references of string column values throughout the spec, not just tables.  However, if that was unclear and you wish to discuss that further, I can pull it out of this PR and issue a separate PR, in favor of getting the other rules approved.  LMK.

Author: In pursuit of progress, I have struck the double-quotes to backticks revisions in this PR.  I will propose that in a separate PR after this one is approved.

Resolving.

* **Consistent Introductory Phrases:** Introductory phrases for normative lists SHOULD be consistent within a section.

> * **Sentence Spacing:** Sentences MUST be delineated with a single space.

Author: ```suggestion
* **Sentence Spacing:** Sentences within paragraphs MUST be delineated with a single space.
```

> * **No Contractions:** In order to maintain a formal, professional tone throughout the specification, contractions MUST NOT be used (e.g., use "do not" instead of "don't").

Author: ```suggestion
* **No Contractions:** Contractions MUST NOT be used within normative requirements (e.g., use "do not" instead of "don't").
```

Author: Let's move this to normative requirements guidelines.

Author: Done!

* **Consistent Structure:** JSON examples SHOULD follow consistent formatting and indentation.

> * **Sentence Spacing**: Sentences between paragraphs MUST be delienated with a single space; double spaces are not permitted.

Author: ```suggestion
* **Sentence Spacing**: Sentences between paragraphs MUST be delineated with a single space; double spaces are not permitted.
```

> * **Sentence Spacing**: Sentences between paragraphs MUST be delienated with a single space; double spaces are not permitted.

Author: Addressed in a separate commit.

* **No Normative Leakage:** Normative keywords (MUST, SHOULD, MAY, etc.) MUST NOT be used in any section other than "Requirements".

> * **Title Case for Subheaders:** Subheaders MUST be denoted in Title Case, with all words capitalized except articles (a, an, the), coordinating conjunctions (and, but, or), and short prepositions (e.g., in, on, at), unless they begin the subheader.

Reviewer 3: just a reminder that we also have the [Header Syntax section](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/guidelines/contributors/markdownpp-guidelines.md#header-syntax) in **MarkdownPP Guidelines for FOCUS Specification**

Author: Add this to the MarkdownPP guidelines.

Author: Done!

>
> 1.0-preview

### Tables

> * **Markdown Table Formatting**: Markdown tables MUST use exactly one space after values, without additional padding or justification to align vertical pipes in the raw markdown.

Author: @jpradocueva opposes this. Believes it should be human readable.

> * **Markdown Table Formatting:** Markdown tables MUST use exactly one space after values, without additional padding or justification to align vertical pipes in the raw markdown.

Author: @jpradocueva opposes this. Believes it should be human readable.

Author: I will take an action to craft this.  Something around less than 120 characters can be padded.

Author: And it becomes SHOULD.

Author: ```suggestion
* **Markdown Table Spacing:**
  * When a markdown table has a maximum row width of less than 120 characters, the table SHOULD be padded with spaces to align the vertical pipes visually.
  * When a markdown table has a maximum row width of 120 characters or more, the table SHOULD use exactly one space after values without additional padding.
```

> * **Markdown Table Spacing**: Markdown tables MUST use only a single space after values. Markdown tables MUST NOT be justified or padded with extra spaces simply to align the vertical pipes visually in the raw markdown.

Author: @jpradocueva opposes this.  Believes it should be human readable.

> * **Markdown Table Spacing**: Markdown tables MUST use only a single space after values. Markdown tables MUST NOT be justified or padded with extra spaces simply to align the vertical pipes visually in the raw markdown.

Author: /gemini I want this to be a single normative requirement.  Give me a different option.

> * **Markdown Table Spacing**: Markdown tables MUST use only a single space after values. Markdown tables MUST NOT be justified or padded with extra spaces simply to align the vertical pipes visually in the raw markdown.

Author: /gemini Can you make it use a single BCP14 statement?

> * **Markdown Table Spacing**: Markdown tables MUST use only a single space after values. Markdown tables MUST NOT be justified or padded with extra spaces simply to align the vertical pipes visually in the raw markdown.

Author: No.  I like this better.

```suggestion
* **Markdown Table Formatting**: Markdown tables MUST use exactly one space after values, without additional padding or justification to align vertical pipes in the raw markdown.
```
